R/get-grades.R

```diff
@@ -169,17 +187,32 @@ lateness <- function(grades_mat, policy_line, category, assignments, weights = c
 #' @rdname score
 #' @export
 drop_n_lowest <- function(grades_mat, policy_line, category, assignments, weights = c()){
-  n <- policy_line #number of drops
-  if (n == 0){ #if no drops
-    return (grades_mat) #no change
+  n <- policy_line # number of drops
+  if (n == 0){ # if no drops
+    return (grades_mat) # no change
   }
+  
+  
   grades_mat[, assignments] <- t(apply(t(grades_mat[, assignments]), 2, function(assign){
-    if (length(assignments) == 1) return (assign) #if only one assignment, no drops
-    lowest_indices <- order(assign)[1:n] #indices of n lowest assignments
-    if (n >= length(assignments)) { #if more drops than assignments
-      lowest_indices <- order(assign)[1:(length(assignments)-1)] #drop all but one
-    }
-    assign[lowest_indices] <- NA #drop designated scores
+    # if only one assignment, no drops
+    # unnecessary but keep for efficiency
+    if (length(assignments) == 1) return (assign) 
+    
+    # count the number of excused assignments
+    num_excused <- sum(is.na(assign))
+    
+    # the number of graded assignments will be length(assignment) - num_excused
+    # a student can drop up to that many - 1. 
+    # so take the min of n and possible allowed drops
+    num_to_drop <- min(c(n, length(assignments) - num_excused - 1))
+ 
+    
+    # if no drops to do return assign
+    if (num_to_drop <= 0) return (assign) 
+    
```

Comment:
It would be negative length(assignments) < num_excused + 1. This would occur if all assignments are excused in a category. Note that reconcile will give a warning in this situation as well.